src/restapi/fIncomingWebhooks.ts

```diff
@@ -55,22 +53,147 @@ export async function fHandleSubcribeNotification(req: Request, res: Response):
+    const promiseEvents: any[] = [];
+
+    // build the new mapped event with promises for where needed
+    for (const event of events) {
+        if (event.field_name === 'ticket_form_id') {
+            promiseEvents.push(getFormNames(zdClient, event));
+            continue;
+        }
+        if (event.field_name === 'assignee_id') {
+            promiseEvents.push(getAssigneeNames(zdClient, event));
+            continue;
+        }
+        if (event.field_name === 'requester_id') {
+            promiseEvents.push(getAssigneeNames(zdClient, event));
+            continue;
+        }
+        if (event.field_name === 'group_id') {
+            promiseEvents.push(getGroupNames(zdClient, event));
+            continue;
+        }
+        mappedArray.push(event);
+    }
+
+    // resolve all the Promises
+    await Promise.all(promiseEvents).then((mappedEvents) => {
+        for (const mappedEvent of mappedEvents) {
+            mappedArray.push(mappedEvent);
+        }
+    });
+    return mappedArray;
+}
+
+async function getFormNames(zdClient: ZDClient, event: any): Promise<any> {
+    const requests: any[] = [];
+    const currReq = zdClient.ticketforms.show(event.value);
+    requests.push(tryPromiseWithMessage(currReq, 'Failed to fetch current ticket form'));
+
+    if (event.previous_value) {
+        const prevReq = zdClient.ticketforms.show(event.previous_value);
+        requests.push(tryPromiseWithMessage(prevReq, 'Failed to fetch previous ticket form'));
+    }
+
+    await Promise.all(requests).then((values) => {
+        event.value = values[0].name;
+        if (values.length === 2) {
+            event.previous_value = values[1].name;
+        }
+    });
+    return event;
+}
+
+async function getGroupNames(zdClient: ZDClient, event: any): Promise<any> {
+    const requests: any[] = [];
+    const currReq = zdClient.groups.show(event.value);
+    requests.push(tryPromiseWithMessage(currReq, 'Failed to fetch current group'));
+
+    if (event.previous_value) {
+        const prevReq = zdClient.groups.show(event.previous_value);
+        requests.push(tryPromiseWithMessage(prevReq, 'Failed to fetch previous group'));
+    }
+
+    await Promise.all(requests).then((values) => {
+        event.value = values[0].name;
+        if (values.length === 2) {
+            event.previous_value = values[1].name;
+        }
+    });
+    return event;
+}
+
+async function getAssigneeNames(zdClient: ZDClient, event: any): Promise<any> {
+    const requests: any[] = [];
+    const currReq = zdClient.users.show(event.value);
+    requests.push(tryPromiseWithMessage(currReq, 'Failed to get current Zendesk user'));
+
+    if (event.previous_value) {
+        const prevReq = zdClient.users.show(event.previous_value);
+        requests.push(tryPromiseWithMessage(prevReq, 'Failed to get previous Zendesk user'));
     }
+
+    await Promise.all(requests).then((values) => {
+        event.value = values[0].name;
+        if (values.length === 2) {
```

Comment:
Lots of duplicated code in these functions. This may or may not make things simpler:

```js
type ZDEvent = {value: any, previous_value?: any};

function processEventValue(zdClientMethod: (value: any) => any, event: ZDEvent): ZDEvent {
    const requests: any[] = [];
    const currReq = zdClientMethod(event.value);
    
    // actually don't know how to compute the correct error message though. Probably needs to also be passed in
    requests.push(tryPromiseWithMessage(currReq, 'Failed to fetch current ticket form'));

    if (event.previous_value) {
        const prevReq = zdClientMethod(event.previous_value);
        requests.push(tryPromiseWithMessage(prevReq, 'Failed to fetch previous ticket form'));
    }

    await Promise.all(requests).then((values) => {
        event.value = values[0].name;
        if (values.length === 2) {
            event.previous_value = values[1].name;
        }
    });
    return event;
}
```

Then in `mapIDsToTextValues`:

```js
if (event.field_name === 'ticket_form_id') {
    promiseEvents.push(processEventValue(zdClient.ticketforms.show, event));
    continue;
}
```

Seems like the code may be more difficult to read this way though. And if anything needs to change on any specific method (may already be the case), then you wouldn't be able to reuse this across the board.

Reply:
ticket created https://mattermost.atlassian.net/browse/MM-35364